pkg/process/monitor/process_monitor.go

```diff
@@ -118,121 +104,107 @@ type ProcessCallback struct {
+// we're iterating the relevant callbacks and trigger them.
+func (pm *ProcessMonitor) handleProcessExit(pid int) {
+	pm.processExitCallbacks.Range(func(key, _ any) bool {
+		callback := key.(*ProcessCallback)
+		pm.handleProcessExitInner(callback, pid)
+		return true
+	})
+
+	pm.runningPids.Delete(pid)
+}
+
+// handleProcessExitInner ensures we should run the given callback on the given PID, and operates if needed.
+// A helper function for handleProcessExit for readability.
+func (pm *ProcessMonitor) handleProcessExitInner(c *ProcessCallback, pid int) {
+	switch c.FilterType {
 	case NAME:
-		metadata, found := pm.runningPids[pid]
+		metadata, found := pm.runningPids.Load(pid)
 		if !found {
 			// we can hit here if a process started before the Exec callback has been registered
 			// and the process Exit, so we don't find his metadata
 			return
 		}
-		pname := metadata.(metadataName).Name
-		if c.Regex.MatchString(pname) {
-			pm.enqueueCallback(c, pid, metadata)
+		pname := metadata.(string)
+		if !c.Regex.MatchString(pname) {
+			return
 		}
+		fallthrough
 	case ANY:
-		pm.enqueueCallback(c, pid, nil)
+		pm.callbackRunner <- func() { c.Callback(pid) }
 	}
 }
 
-// terminateProcessMonitor is a helper function of Initialize. The goal is to make sure we properly terminate our
-// go-routine.
-func (pm *ProcessMonitor) terminateProcessMonitor() {
-	log.Info("netlink process monitor ended")
-	pm.wg.Done()
-	close(pm.callbackRunner)
-}
-
-// Initialize will scan all running processes and execute matching callbacks
-// Once it's done all new events from netlink socket will be processed by the main async loop
-func (pm *ProcessMonitor) Initialize() error {
-	pm.m.Lock()
-	defer pm.m.Unlock()
-
-	pm.refcount++
-	if pm.isInitialized {
-		return nil
-	}
-
-	pm.events = make(chan netlink.ProcEvent, processMonitorMaxEvents)
-	pm.done = make(chan struct{})
-	pm.errors = make(chan error, 10)
-
-	hostProc := util.HostProc()
-	if err := util.WithRootNS(hostProc, func() error {
-		return netlink.ProcEventMonitor(pm.events, pm.done, pm.errors)
+// initNetlinkProcessEventMonitor initialize the netlink socket filter for process event monitor.
+func (pm *ProcessMonitor) initNetlinkProcessEventMonitor() error {
+	pm.netlinkDoneChannel = make(chan struct{})
+	pm.netlinkErrorsChannel = make(chan error, 10)
+	pm.netlinkEventsChannel = make(chan netlink.ProcEvent, processMonitorMaxEvents)
 
+	if err := util.WithRootNS(util.GetProcRoot(), func() error {
+		return netlink.ProcEventMonitor(pm.netlinkEventsChannel, pm.netlinkDoneChannel, pm.netlinkErrorsChannel)
 	}); err != nil {
 		return fmt.Errorf("couldn't initialize process monitor: %s", err)
 	}
 
-	pm.callbackRunner = make(chan func(), runtime.NumCPU())
-	for i := 0; i < runtime.NumCPU(); i++ {
-		pm.wg.Add(1)
+	return nil
+}
+
+// initCallbackRunner runs multiple workers that run tasks sent over a queue.
+func (pm *ProcessMonitor) initCallbackRunner() {
+	cpuNum := runtime.NumCPU()
```

Comment:
I _believe_ `runtime.NumCPU` is not aware of cgroups/CPU quotas so I'd avoid this.
I think there are some libraries that deal with that (eg https://github.com/uber-go/automaxprocs) and we might even have something like that in the agent code-base, but we certainly don't want to create 96 go-routines on a 96 vCPU node that has 1 core allocated to system-probe.

Reply:
goroutines are not threads
so having 96 go routines does not mean that we're utilizing 96 cores

Reply:
using internal runtime.NumVCPU

Reply:
> goroutines are not threads
I didn't mean that.

My point is that if you have 96 Go routines doing CPU-bound work on a process that have access to 1 core, you're likely only making things worse by adding synchronization overhead, because 95 Go routines will be idle while only 1 will be working at given point in time.

What you likely want is to have more or less the same number of go routines as go "processes" because that is as much parallelism you'll be able to get. If your workload involves blocking operations (like IO) then yeah, you can definitely benefit from having _more_ go routines than go processes/CPU cores because then you can have multiple things being executed _concurrently_, but I guarantee you that not in a ratio of 96-to-1 😅

Reply:
thanks for the clarification!